packages/editor/src/lib/exports/StyleEmbedder.ts

```diff
@@ -0,0 +1,310 @@
+	) {
+		const defaultStyles = shouldRespectDefaults
+			? getDefaultStylesForTagName(element.tagName.toLowerCase())
+			: NO_STYLES
+
+		const parentStyles = shouldSkipInheritedParentStyles
+			? this.styles.get(element.parentElement as Element)?.self ?? NO_STYLES
+			: NO_STYLES
+
+		const info: ElementStyleInfo = {
+			self: styleFromElement(element, { defaultStyles, parentStyles }),
+			before: styleFromPseudoElement(element, '::before'),
+			after: styleFromPseudoElement(element, '::after'),
+		}
+		this.styles.set(element, info)
+	}
+
+	fetchResources() {
+		const promises: Promise<void>[] = []
+
+		for (const info of this.styles.values()) {
+			for (const styles of objectMapValues(info)) {
+				if (!styles) continue
+				for (const [property, value] of Object.entries(styles)) {
+					if (!value) continue
+					if (property === 'font-family') {
+						this.fonts.onFontFamilyValue(value)
+					}
+
+					const urlMatches = parseCssValueUrls(value)
+					if (urlMatches.length === 0) continue
+
+					promises.push(
+						...urlMatches.map(async ({ url, original }) => {
+							const dataUrl = (await resourceToDataUrl(url)) ?? 'data:'
+							styles[property] = value.replace(original, `url("${dataUrl}")`)
+						})
+					)
+				}
+			}
+		}
+
+		return Promise.all(promises)
+	}
+
+	// custom elements are tricky. if we serialize the dom as-is, the custom elements wont have
+	// their shadow-dom contents serialized. after we've read all the styles, we need to unwrap the
+	// contents of each custom elements shadow dom directly into the parent element itself.
+	unwrapCustomElements() {
+		const visited = new Set<Node>()
+
+		const visit = (element: Element, clonedParent: Element | null) => {
+			if (visited.has(element)) return
+			visited.add(element)
+
+			const shadowRoot = element.shadowRoot
+
+			// if (shadowRoot || clonedParent) debugger
+
+			if (shadowRoot) {
+				const clonedCustomEl = document.createElement('div')
+				this.styles.set(clonedCustomEl, this.styles.get(element)!)
+
+				clonedCustomEl.setAttribute('data-tl-custom-element', element.tagName)
+				;(clonedParent ?? element.parentElement!).appendChild(clonedCustomEl)
+
+				for (const child of shadowRoot.childNodes) {
+					if (child instanceof Element) {
+						visit(child, clonedCustomEl)
+					} else {
+						clonedCustomEl.appendChild(child.cloneNode(true))
+					}
+				}
+
+				element.remove()
+			} else if (clonedParent) {
+				if (element.tagName.toLowerCase() === 'style') {
+					// we don't clone style tags at that would break the style scoping. instead we rely on the computed styles we've already read
+					return
+				}
+
+				const clonedEl = element.cloneNode(false) as Element
+				this.styles.set(clonedEl, this.styles.get(element)!)
+
+				clonedParent.appendChild(clonedEl)
+
+				for (const child of getRenderedChildNodes(element)) {
+					if (child instanceof Element) {
+						visit(child, clonedEl)
```

Comment:
this might be worth having a max-depth check here somewhere. i'm worried about an infinite loop potentially...

Reply:
Is there a way we could have a loop that wouldn't be prevented by the `visited` set?

Reply:
i think the `visited` set suffices, it's just a paranoia thing. i guess seeing the `cloneNode`+recursion makes me think it'll be generating accidentally more nodes and then continue consuming them ad infinitum. 

the browser will throw anyway if we hit recursion so it's prbly fine